File: iOS/Objc/Examples/cases/ObjectACL.m
#import <XCTest/XCTest.h>
#import <QCloudCOSXML/QCloudCOSXML.h>
#import <QCloudCOSXML/QCloudUploadPartRequest.h>
#import <QCloudCOSXML/QCloudCompleteMultipartUploadRequest.h>
#import <QCloudCOSXML/QCloudAbortMultipfartUploadRequest.h>
#import <QCloudCOSXML/QCloudMultipartInfo.h>
#import <QCloudCOSXML/QCloudCompleteMultipartUploadInfo.h>


@interface ObjectACL : XCTestCase <QCloudSignatureProvider, QCloudCredentailFenceQueueDelegate>

@property (nonatomic) QCloudCredentailFenceQueue* credentialFenceQueue;

@end

@implementation ObjectACL

- (void)setUp {
    // 注册默认的 COS 服务
    QCloudServiceConfiguration* configuration = [QCloudServiceConfiguration new];
    configuration.appID = @"1250000000";
    configuration.signatureProvider = self;
    QCloudCOSXMLEndPoint* endpoint = [[QCloudCOSXMLEndPoint alloc] init];
    endpoint.regionName = @"ap-guangzhou";//服务地域名称，可用的地域请参考注释
    configuration.endpoint = endpoint;
    [QCloudCOSXMLService registerDefaultCOSXMLWithConfiguration:configuration];
    [QCloudCOSTransferMangerService registerDefaultCOSTransferMangerWithConfiguration:configuration];

    // 脚手架用于获取临时密钥
    self.credentialFenceQueue = [QCloudCredentailFenceQueue new];
    self.credentialFenceQueue.delegate = self;
}

- (void) fenceQueue:(QCloudCredentailFenceQueue * )queue requestCreatorWithContinue:(QCloudCredentailFenceQueueContinue)continueBlock
{
    QCloudCredential* credential = [QCloudCredential new];
    //在这里可以同步过程从服务器获取临时签名需要的 secretID，secretKey，expiretionDate 和 token 参数
    credential.secretID = @"COS_SECRETID";
    credential.secretKey = @"COS_SECRETKEY";
    credential.token = @"COS_TOKEN";
    /*强烈建议返回服务器时间作为签名的开始时间，用来避免由于用户手机本地时间偏差过大导致的签名不正确 */
    credential.startDate = [[[NSDateFormatter alloc] init] dateFromString:@"startTime"]; // 单位是秒
    credential.experationDate = [[[NSDateFormatter alloc] init] dateFromString:@"expiredTime"];
    QCloudAuthentationV5Creator* creator = [[QCloudAuthentationV5Creator alloc]
        initWithCredential:credential];
    continueBlock(creator, nil);
}

- (void) signatureWithFields:(QCloudSignatureFields*)fileds
                     request:(QCloudBizHTTPRequest*)request
                  urlRequest:(NSMutableURLRequest*)urlRequst
                   compelete:(QCloudHTTPAuthentationContinueBlock)continueBlock
{
    [self.credentialFenceQueue performAction:^(QCloudAuthentationCreator *creator,
                                               NSError *error) {
        if (error) {
            continueBlock(nil, error);
        } else {
            QCloudSignature* signature =  [creator signatureForData:urlRequst];
            continueBlock(signature, nil);
        }
    }];
}

/**
 * 设置对象访问权限控制列表
 */
- (void)putObjectAcl {

    //.cssg-snippet-body-start:[objc-put-object-acl]
    QCloudPutObjectACLRequest* request = [QCloudPutObjectACLRequest new];
    
    // 对象键，是对象在 COS 上的完整路径，如果带目录的话，格式为 "dir1/object1"
    request.object = @"exampleobject";
    
    // 存储桶名称，格式为 BucketName-APPID
    request.bucket = @"examplebucket-1250000000";
    
    NSString *grantString = [NSString stringWithFormat:@"id=\"%@\"",@"100000000001"];
    
    // grantFullControl 等价于 grantRead + grantWrite
    // 赋予被授权者读写权限。
    request.grantFullControl = grantString;
    // 赋予被授权者读权限。
    request.grantRead = grantString;
    // 赋予被授权者写权限。
    request.grantWrite = grantString;
    
    [request setFinishBlock:^(id outputObject, NSError *error) {
        // 可以从 outputObject 中获取 response 中 etag 或者自定义头部等信息
        NSDictionary* info = (NSDictionary *) outputObject;
    }];
    
    [[QCloudCOSXMLService defaultCOSXML] PutObjectACL:request];
    
    //.cssg-snippet-body-end
}

/**
 * 查询对象访问权限控制列表
 */
- (void)getObjectAcl {

    //.cssg-snippet-body-start:[objc-get-object-acl]
    QCloudGetObjectACLRequest *request = [QCloudGetObjectACLRequest new];
    
    // 对象键，是对象在 COS 上的完整路径，如果带目录的话，格式为 "dir1/object1"
    request.object = @"exampleobject";
    
    // 存储桶名称，格式为 BucketName-APPID
    request.bucket = @"examplebucket-1250000000";
    
    __block QCloudACLPolicy* policy;
    [request setFinishBlock:^(QCloudACLPolicy * _Nonnull result,
                              NSError * _Nonnull error) {
        
        policy = result;
        // result.accessControlList; 被授权者与权限的信息
        // result.owner; 持有者的信息
    }];
    
    [[QCloudCOSXMLService defaultCOSXML] GetObjectACL:request];
    
    //.cssg-snippet-body-end

}
// .cssg-methods-pragma


- (void)testObjectACL {
    // 设置对象 ACL
    [self putObjectAcl];
        
    // 获取对象 ACL
    [self getObjectAcl];
    // .cssg-methods-pragma
        
}

@end

File: iOS/Swift/Examples/cases/ObjectACL.swift
import XCTest
import QCloudCOSXML

class ObjectACL: XCTestCase,QCloudSignatureProvider,QCloudCredentailFenceQueueDelegate{
    
    var credentialFenceQueue:QCloudCredentailFenceQueue?;
    
    override func setUp() {
        let config = QCloudServiceConfiguration.init();
        config.signatureProvider = self;
        config.appID = "1253653367";
        let endpoint = QCloudCOSXMLEndPoint.init();
        endpoint.regionName = "ap-guangzhou";//服务地域名称，可用的地域请参考注释
        endpoint.useHTTPS = true;
        config.endpoint = endpoint;
        QCloudCOSXMLService.registerDefaultCOSXML(with: config);
        QCloudCOSTransferMangerService.registerDefaultCOSTransferManger(with: config);
        
        // 脚手架用于获取临时密钥
        self.credentialFenceQueue = QCloudCredentailFenceQueue();
        self.credentialFenceQueue?.delegate = self;
    }
    
    func fenceQueue(_ queue: QCloudCredentailFenceQueue!,
                    requestCreatorWithContinue continueBlock: QCloudCredentailFenceQueueContinue!) {
        let cre = QCloudCredential.init();
        //在这里可以同步过程从服务器获取临时签名需要的 secretID，secretKey，expiretionDate 和 token 参数
        cre.secretID = "COS_SECRETID";
        cre.secretKey = "COS_SECRETKEY";
        cre.token = "COS_TOKEN";
        /*强烈建议返回服务器时间作为签名的开始时间，用来避免由于用户手机本地时间偏差过大导致的签名不正确 */
        cre.startDate = DateFormatter().date(from: "startTime"); // 单位是秒
        cre.experationDate = DateFormatter().date(from: "expiredTime");
        let auth = QCloudAuthentationV5Creator.init(credential: cre);
        continueBlock(auth,nil);
    }
    
    func signature(with fileds: QCloudSignatureFields!,
                   request: QCloudBizHTTPRequest!,
                   urlRequest urlRequst: NSMutableURLRequest!,
                   compelete continueBlock: QCloudHTTPAuthentationContinueBlock!) {
        self.credentialFenceQueue?.performAction({ (creator, error) in
            if error != nil {
                continueBlock(nil,error!);
            }else{
                let signature = creator?.signature(forData: urlRequst);
                continueBlock(signature,nil);
            }
        })
    }
    
    
    // 设置对象 ACL
    func putObjectAcl() {
        //.cssg-snippet-body-start:[swift-put-object-acl]
        let putObjectACl = QCloudPutObjectACLRequest.init();
        
        // 存储桶名称，格式为 BucketName-APPID
        putObjectACl.bucket = "examplebucket-1250000000";
        
        // 对象键，是对象在 COS 上的完整路径，如果带目录的话，格式为 "dir1/object1"
        putObjectACl.object = "exampleobject";
        let grantString = "id=\"100000000001\"";
        
        // grantFullControl 等价于 grantRead + grantWrite
        putObjectACl.grantFullControl = grantString;
        // 赋予被授权者读权限。
        putObjectACl.grantRead = grantString;
        // 赋予被授权者写权限。
        putObjectACl.grantWrite = grantString;
        
        putObjectACl.finishBlock = {(result,error)in
            if let result = result {
                // result 包含响应的 header 信息
            } else {
                print(error!);
            }
        }
        QCloudCOSXMLService.defaultCOSXML().putObjectACL(putObjectACl);
        
        //.cssg-snippet-body-end
    }
    
    
    // 获取对象 ACL
    func getObjectAcl() {
        //.cssg-snippet-body-start:[swift-get-object-acl]
        let getObjectACL = QCloudGetObjectACLRequest.init();
        
        // 存储桶名称，格式为 BucketName-APPID
        getObjectACL.bucket = "examplebucket-1250000000";
        
        // 对象键，是对象在 COS 上的完整路径，如果带目录的话，格式为 "dir1/object1"
        getObjectACL.object = "exampleobject";
        getObjectACL.setFinish { (result, error) in
            if let result = result {
                // 对象授权信息
                let acl = result.accessControlList
            } else {
                print(error!);
            }
        }
        QCloudCOSXMLService.defaultCOSXML().getObjectACL(getObjectACL);
        
        //.cssg-snippet-body-end
    }
    // .cssg-methods-pragma
    
    
    func testObjectACL() {
        // 设置对象 ACL
        self.putObjectAcl();
        // 获取对象 ACL
        self.getObjectAcl();
        // .cssg-methods-pragma
    }
}
